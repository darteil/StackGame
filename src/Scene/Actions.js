export const ADD_RECORD = 'ADD_RECORD';
export const SET_TOP_RECORD = 'SET_TOP_RECORD';

export function addRecord(record) {
  return {
    type: ADD_RECORD,
    record
  };
}

export function setTopRecord(count) {
  return {
    type: SET_TOP_RECORD,
    count
  };
}
