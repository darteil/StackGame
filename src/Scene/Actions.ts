import { ADD_RECORD, SET_TOP_RECORD, TOGGLE_UI, IToggleUI, IAddRecord, ISetTopRecord, Record } from './types';

export function addRecord(record: Record): IAddRecord {
  return {
    type: ADD_RECORD,
    record
  };
}

export function setTopRecord(count: number, heightStack: number): ISetTopRecord {
  return {
    type: SET_TOP_RECORD,
    count,
    heightStack
  };
}

export function toggleUI(): IToggleUI {
  return {
    type: TOGGLE_UI
  };
}
