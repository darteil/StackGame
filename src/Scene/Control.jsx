import React from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';
import styles from './styles.css';

const Control = props => (
  <div className={styles.control}>
    <button className={ClassNames(styles['fuller-button'], styles.blue)} onClick={props.restartGame}>
      Restart
    </button>
  </div>
);

Control.propTypes = {
  restartGame: PropTypes.func.isRequired
};

export default Control;
