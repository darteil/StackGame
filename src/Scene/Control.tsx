import React from 'react';
import PropTypes from 'prop-types';
import styles from './styles.css';

interface IProps {
  restartGame: () => void;
}

const Control = (props: IProps) => (
  <div className={styles.control}>
    <button className={styles['restart-button']} onClick={props.restartGame}>
      Restart
    </button>
  </div>
);

Control.propTypes = {
  restartGame: PropTypes.func.isRequired
};

export default Control;
