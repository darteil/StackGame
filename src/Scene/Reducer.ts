import { IGameDataState, RecordActionTypes, ADD_RECORD, SET_TOP_RECORD, TOGGLE_UI } from './types';

const initialState: IGameDataState = {
  versionData: 1,
  listOfRecords: [],
  topRecord: 0,
  topHeightStack: 0,
  UI: true
};

export default function GameData(state = initialState, action: RecordActionTypes): IGameDataState {
  switch (action.type) {
    case ADD_RECORD: {
      const newListOfRecords = [...state.listOfRecords];
      newListOfRecords.push(action.record);

      return { ...state, listOfRecords: newListOfRecords };
    }
    case SET_TOP_RECORD: {
      return { ...state, topRecord: action.count, topHeightStack: action.heightStack };
    }
    case TOGGLE_UI: {
      return { ...state, UI: !state.UI };
    }
    default:
      return state;
  }
}
