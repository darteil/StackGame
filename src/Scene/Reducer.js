import { ADD_RECORD, SET_TOP_RECORD } from './Actions';

const initialState = {
  listOfRecords: [],
  topRecord: 0
};

export default function GameData(state = initialState, action) {
  switch (action.type) {
    case ADD_RECORD: {
      const newListOfRecords = [...state.listOfRecords];
      const record = { ...action.record };

      record.id = newListOfRecords.length + 1;

      newListOfRecords.push(record);

      return { ...state, listOfRecords: newListOfRecords };
    }
    case SET_TOP_RECORD: {
      return { ...state, topRecord: action.count };
    }
    default:
      return state;
  }
}
