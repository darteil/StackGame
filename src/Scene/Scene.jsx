/* global document */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import moment from 'moment-timezone';
import { showMessage } from 'App/Message/Actions';
import { addRecord, setTopRecord } from './Actions';
import Game from '../Game/index';
import Control from './Control';
import styles from './styles.css';


class Scene extends Component {
  constructor(props) {
    super(props);

    this.state = {
      gameStatus: 'stopped', // launched or stopped
      count: 0
    };

    this.timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;

    this.container = React.createRef();
    this.game = new Game(this.container);
    this.startGame = this.startGame.bind(this);
    this.stopGame = this.stopGame.bind(this);
    this.restartGame = this.restartGame.bind(this);
    this.setNewStack = this.setNewStack.bind(this);
  }

  componentDidMount() {
    document.addEventListener('keydown', this.setNewStack);
    this.game.init();
    this.startGame();
    if (this.props.listOfRecords.length === 0) {
      this.props.showMessage('Press "w" to set block =)');
    }
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.setNewStack);
  }

  setNewStack(event) {
    if (this.state.gameStatus === 'stopped' || event.which !== 87) {
      return false;
    }

    this.game.setNewStack();
    this.setState(() => ({ count: this.game.getCount() }));

    if (this.game.getStopStatusGame()) {
      this.setState(() => ({ gameStatus: 'stopped' }));

      if (this.props.topRecord !== 0) {
        if (this.state.count > this.props.topRecord) {
          this.props.setTopRecord(this.state.count);
          this.props.showMessage('New record!');
        }
      } else {
        this.props.setTopRecord(this.state.count);
      }

      this.props.addRecord({ time: moment().tz(this.timeZone).format('MMMM Do YYYY, HH:mm:ss'), count: this.state.count });
    }

    return true;
  }

  startGame() {
    if (this.state.gameStatus === 'launched') {
      return false;
    }
    this.game.start();
    this.setState(() => ({ gameStatus: 'launched' }));
    return true;
  }

  stopGame() {
    this.game.stopGame();
  }

  restartGame() {
    this.setState(() => ({
      count: 0,
      gameStatus: 'launched'
    }));
    this.game.restartGame();
  }

  render() {
    return (
      <div className={styles['scene-wrap']}>
        <Control restartGame={this.restartGame} />
        <div className={styles.count}>{this.state.count}</div>
        <div className={styles.scene} ref={this.container} />
      </div>
    );
  }
}

Scene.propTypes = {
  addRecord: PropTypes.func.isRequired,
  setTopRecord: PropTypes.func.isRequired,
  showMessage: PropTypes.func.isRequired,
  topRecord: PropTypes.number.isRequired,
  listOfRecords: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
  topRecord: state.GameData.topRecord,
  listOfRecords: state.GameData.listOfRecords
});

const mapDispatchToProps = dispatch => ({
  addRecord: (record) => {
    dispatch(addRecord(record));
  },
  setTopRecord: (record) => {
    dispatch(setTopRecord(record));
  },
  showMessage: (text) => {
    dispatch(showMessage(text));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Scene);

