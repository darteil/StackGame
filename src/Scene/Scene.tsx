/* global document */
import React, { Component } from 'react';
import moment from 'moment-timezone';
import { Record } from '../ListOfRecords/types';
import { v1 as uuid } from 'uuid';
import Game from '../Game';
import Control from './Control';
import styles from './styles.css';

interface IProps {
  addRecord: (record: Record) => void;
  setTopRecord: (count: number, height: number) => void;
  showMessage: (text: string) => void;
  topRecord: number;
  listOfRecords: Record[];
}

interface IState {
  gameStatus: string;
  count: number;
}

export default class Scene extends Component<IProps, IState> {
  private timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;
  private container!: React.RefObject<HTMLDivElement>;
  private game: any;

  constructor(props: IProps) {
    super(props);

    this.state = {
      gameStatus: 'stopped', // launched or stopped
      count: 0
    };

    this.game = null;
    this.container = React.createRef();
    this.setNewStack = this.setNewStack.bind(this);
  }

  componentDidMount() {
    this.game = new Game(this.container.current as HTMLElement);
    this.game.init();
    this.startGame();
    // this.game.enableOrbitControls();
    // this.game.enableAxesHelper();

    window.addEventListener('keydown', this.setNewStack);
    if (this.props.listOfRecords.length === 0) {
      this.props.showMessage('Press "w" to set block =)');
    }
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this.setNewStack);
  }

  setNewStack = (event: KeyboardEvent) => {
    if (this.state.gameStatus === 'stopped' || event.which !== 87) {
      return;
    }

    this.game.setNewStack();
    this.setState(() => ({ count: this.game.getCount() }));

    if (this.game.getStopStatusGame()) {
      this.setState(() => ({ gameStatus: 'stopped' }));

      if (this.props.topRecord !== 0) {
        if (this.state.count > this.props.topRecord) {
          this.props.setTopRecord(this.state.count, this.game.getStackHeight());
          this.props.showMessage('New record!');
        }
      } else {
        this.props.setTopRecord(this.state.count, this.game.getStackHeight());
      }

      this.props.addRecord({
        id: uuid(),
        time: moment()
          .tz(this.timeZone)
          .format('MMMM Do YYYY, HH:mm:ss'),
        count: this.state.count,
        heightStack: this.game.getStackHeight()
      });
    }

    return;
  };

  startGame = () => {
    if (this.state.gameStatus === 'launched') {
      return false;
    }
    this.game.start();
    this.setState(() => ({ gameStatus: 'launched' }));
    return true;
  };

  stopGame = () => {
    this.game.stopGame();
  };

  restartGame = () => {
    this.setState(() => ({
      count: 0,
      gameStatus: 'launched'
    }));
    this.game.restartGame();
  };

  render() {
    return (
      <div className={styles['scene-wrap']}>
        <Control restartGame={this.restartGame} />
        <div className={styles.count}>{this.state.count}</div>
        <div className={styles.scene} ref={this.container} />
      </div>
    );
  }
}
