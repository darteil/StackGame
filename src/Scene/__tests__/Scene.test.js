import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Scene from '../Scene';

describe('Scene component', () => {
  const props = {
    listOfRecords: [
      {
        id: 1,
        time: 'September 24th 2018, 16:15:26',
        count: 3,
        heightStack: 3
      },
      {
        id: 2,
        time: 'September 29th 2018, 16:11:32',
        count: 16,
        heightStack: 17
      }
    ],
    topRecord: 16,
    showMessage: jest.fn(),
    setTopRecord: jest.fn(),
    addRecord: jest.fn()
  };

  const output = shallow(<Scene {...props} />, { disableLifecycleMethods: true });

  it('should render correctly', () => {
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});
