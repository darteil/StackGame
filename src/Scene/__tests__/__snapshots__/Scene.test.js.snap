// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene component should render correctly 1`] = `
<div
  className="scene-wrap"
>
  <Control
    restartGame={[Function]}
  />
  <div
    className="count"
  >
    0
  </div>
  <div
    className="scene"
  />
</div>
`;
